kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java

```diff
@@ -175,16 +186,101 @@ public KieServerState load(@NotNull String serverId) {
     
     @Override
     public boolean isKieServerReady() {
-        if (isKieServerRuntime()) {
-            return KieServerLocator.getInstance().isKieServerReady();
+        return KieServerLocator.getInstance().isKieServerReady();
+    }
+
+    protected ConfigMap createNewKieServerStateCM(KieServerState kieServerState, String serverId, OpenShiftClient client) {
+        String stateXML = xs.toXML(kieServerState);
+        Map<String, String> labels = null;
+        if (Boolean.parseBoolean(kieServerState.getConfiguration()
+                                 .getConfigItemValue(KieServerConstants.KIE_SERVER_STATE_IMMUTABLE, "false"))) {
+            labels = Collections.singletonMap(CFG_MAP_LABEL_NAME, CFG_MAP_LABEL_VALUE_IMMUTABLE);
+        } else {    
+            labels = Collections.singletonMap(CFG_MAP_LABEL_NAME, CFG_MAP_LABEL_VALUE_USED);
+        }
+        return client.configMaps().createOrReplace(new ConfigMapBuilder()
+                                   .withNewMetadata()
+                                     .withName(serverId)
+                                     .withLabels(labels)
+                                   .endMetadata()
+                                   .withData(Collections.singletonMap(CFG_MAP_DATA_KEY, stateXML))
+                                   .build());
+    }
+
+    protected boolean isServerStateUpdateAllowed(DeploymentConfig dc, ConfigMap cm, KieServerState newState) {
+        if (cm == null) {
+            throw new IllegalStateException("KieServerState ConfigMap must exist before update.");
+        }
+        if (cm.getMetadata().getLabels() == null) {
+            return false;
+        }
+        if (cm.getMetadata().getLabels().containsValue(CFG_MAP_LABEL_VALUE_USED)) {
+            if (isKieServerRuntime()) {
+                return isUpdateByKieServerProcessAllowed(dc, cm, newState);
+            } else {
+                return isUpdateByNonKieServerProcessAllowed(dc);
+            }
+        }
+        if (cm.getMetadata().getLabels().containsValue(CFG_MAP_LABEL_VALUE_IMMUTABLE)) {
+            logger.warn("Add or remove kie container not allowed for immutable kie server: {}",
+                        cm.getMetadata().getName());
+        }
+        return false;
+    }
+    
+    /**
+     * To be compatible with non kieserver process, (workbench), kie container at certain status,
+     * i.e. STOPPED, should not be removed.
+     * @param dc
+     * @param c
+     * @param newState
+     * @return
+     */
+    protected boolean isKieContainerRemovalAllowed(ConfigMap cm, KieServerState newState) {
+        KieServerState state = (KieServerState) xs.fromXML(cm.getData().get(CFG_MAP_DATA_KEY));
+        for (KieContainerResource container : state.getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.STOPPED) &&
+                newState.getContainers().stream()
+                        .filter(c -> c.getContainerId().equals(container.getContainerId()))
```

Comment:
Could be replaced by .noneMatch(c -> c.getContainerId().equals(container.getContainerId())))

Reply:
Updated as per your suggestion